File: ProjectScrapeData/ScrapeData/ScrapeDataDOJI.js
const axios = require('axios');
const cheerio = require('cheerio');
const https = require('https');
const fs = require('fs');
const { parse } = require('json2csv');
const { exec } = require('child_process');

async function fetchGoldTables() {
    try {
        const dateNow = new Date();
        const formatDate = `${dateNow.getDate()}-${dateNow.getMonth() + 1}-${dateNow.getFullYear()}`
        console.log(`Lấy dữ liệu vào ${formatDate}`);

        const url = 'https://doji.vn/bang-gia-vang/';
        const { data } = await axios.get(url, {
            httpsAgent: new https.Agent({
                rejectUnauthorized: false // Bỏ qua xác minh SSL
            })
        });
        const $ = cheerio.load(data);

        const records = [];
        let idCounter = 1; // Biến đếm id cho từng hàng

        $('div._table').each((_, table) => {
            const nameBrach = $(table).find('div._title-tax').first().text().replace(/giá vàng trang sức tại|giá vàng/, '').trim();
            const goldTypes = [];
            $(table).find('div._taxonomy ._block').each((_, element) => {
                const type = $(element).text().trim();
                goldTypes.push(type);
            });

            const buyPrices = [];
            $(table).find('div._buy ._block').each((_, element) => {
                const price = $(element).text().trim() == '-' ? ' ' : $(element).text().trim() + ',000';
                buyPrices.push(price);
            });

            const sellPrices = [];
            $(table).find('div._Sell ._block').each((_, element) => {
                const price = $(element).text().trim() == '-' ? ' ' : $(element).text().trim() + ',000';
                sellPrices.push(price);
            });

            goldTypes.forEach((type, i) => {
                if (type) {
                    records.push({
                        id: idCounter++,
                        goldTypes: type,
                        nameBrach,
                        buy: buyPrices[i] || 'N/A',
                        sell: sellPrices[i] || 'N/A'
                    });
                }
            });
        });

        const csvFilePath = `D:\\Datawarehouse\\FileDataCSV\\gold_prices_doji_${formatDate}.csv`;
        const csvRecords = records.map(record => ({
            Id: record.id,
            TypeName: record.goldTypes,
            Company: record.nameBrach,
            BuyValue: record.buy,
            SellValue: record.sell
        }));
        const csv = parse(csvRecords, { fields: ['Id', 'TypeName', 'Company', 'BuyValue', 'SellValue'] });
        fs.writeFileSync(csvFilePath, csv);

        const xlsxFilePath = `D:\\Datawarehouse\\FileDataXLSX\\gold_prices_doji_${formatDate}.xlsx`;
        const methodConvert = 'D:\\Datawarehouse\\convertCSVtoXLSX.js'
        exec(`node ${methodConvert} ${csvFilePath} ${xlsxFilePath}`, (error, stdout, stderr) => {
            if (error) {
                console.error(`Error converting to XLSX: ${error.message}`);
                return;
            }
            if (stderr) {
                console.error(`stderr: ${stderr}`);
                return;
            }
            console.log(stdout);
        });

        console.log('Dữ liệu đã được ghi vào file goldPrices.csv');
    } catch (error) {
        console.error('Có lỗi xảy ra:', error);
    }
}

fetchGoldTables();

File: scrapeDatafromWeb/src/main/java/com/example/scrapeDatafromWeb/service/ScrapeDataDOJI.java
package com.example.scrapeDatafromWeb.service;

import com.example.scrapeDatafromWeb.entity.FileConfigEntity;
import com.example.scrapeDatafromWeb.repository.FileConfigRepository;
import com.example.scrapeDatafromWeb.request.FileLogRequest;
import com.example.scrapeDatafromWeb.utils.SSLUtils;
import org.apache.poi.ss.usermodel.*;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.stereotype.Service;

import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;


@Component
@Service
public class ScrapeDataDOJI {

    @Autowired
    private FileLogService fileLogService;

    @Autowired
    private FileConfigRepository fileConfigRepository;

    private FileConfigEntity formatName(){
        FileConfigEntity entity = fileConfigRepository.findByNameAndEventAndStatus("gold_prices_doji_", "scrape", "active");

        return entity;
    }

    public void scrapeGoldPrices() {
        String url = "https://doji.vn/bang-gia-vang/";
//        String csvDirectory = "../scrapeCSV/";
        String xlsxDirectory = "../changeCSVtoXLSX/";

        try {
            // Fetch the webpage
            SSLUtils.disableSSLCertificateChecking();
            Document document = Jsoup.connect(url).timeout(5000).get();

            // Prepare date and file names
            String dateNow = new SimpleDateFormat("dd-MM-yyyy").format(new Date());
//            String csvFilePath = csvDirectory + "gold_prices_doji_" + dateNow + ".csv";
            String csvFilePath = formatName().getPath() + dateNow + ".csv";
            String xlsxFilePath = xlsxDirectory + "gold_prices_doji_" + dateNow + ".xlsx";

            // Extract data
            List<GoldPriceRecordDOJI> records = new ArrayList<>();
            int idCounter = 1;

            Elements tables = document.select("div._table");
            for (Element table : tables) {
                String nameBranch = table.selectFirst("div._title-tax").text()
                        .replace("giá vàng trang sức tại", "")
                        .replace("giá vàng", "").trim();

                Elements goldTypes = table.select("div._taxonomy ._block");
                Elements buyPrices = table.select("div._buy ._block");
                Elements sellPrices = table.select("div._Sell ._block");

                for (int i = 0; i < goldTypes.size(); i++) {
                    String type = goldTypes.get(i).text().trim();
                    String buy = i < buyPrices.size() ? buyPrices.get(i).text().trim() + ",000" : "N/A";
                    String sell = i < sellPrices.size() ? sellPrices.get(i).text().trim() + ",000" : "N/A";

                    records.add(new GoldPriceRecordDOJI(idCounter++, type, nameBranch, buy, sell));
                }
            }

            // Save to CSV
            writeCsvFile(records, csvFilePath);

            // Convert to XLSX
            writeXlsxFile(records, xlsxFilePath);

            System.out.println("Dữ liệu đã được lưu thành công!");
            System.out.println("CSV: " + csvFilePath);
            FileLogRequest request = new FileLogRequest(csvFilePath, "scraper", "success");
            fileLogService.saveFileLogs(request);

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static void writeCsvFile(List<GoldPriceRecordDOJI> records, String filePath) throws IOException {
        try (FileWriter writer = new FileWriter(filePath)) {
            writer.append("Id,TypeName,Company,BuyValue,SellValue\n");
            for (GoldPriceRecordDOJI record : records) {
                writer.append(record.toCsvLine()).append("\n");
            }
        }
    }

    private static void writeXlsxFile(List<GoldPriceRecordDOJI> records, String filePath) throws IOException {
        Workbook workbook = new XSSFWorkbook();
        Sheet sheet = workbook.createSheet("Gold Prices");

        // Header row
        Row headerRow = sheet.createRow(0);
        String[] headers = {"Id", "TypeName", "Company", "BuyValue", "SellValue"};
        for (int i = 0; i < headers.length; i++) {
            Cell cell = headerRow.createCell(i);
            cell.setCellValue(headers[i]);
        }

        // Data rows
        int rowIndex = 1;
        for (GoldPriceRecordDOJI record : records) {
            Row row = sheet.createRow(rowIndex++);
            row.createCell(0).setCellValue(record.getId());
            row.createCell(1).setCellValue(record.getTypeName());
            row.createCell(2).setCellValue(record.getCompany());
            row.createCell(3).setCellValue(record.getBuyValue());
            row.createCell(4).setCellValue(record.getSellValue());
        }

        // Write to file
        Path path = Paths.get(filePath);
        Files.createDirectories(path.getParent());
        workbook.write(Files.newOutputStream(path));
        workbook.close();
    }
}

class GoldPriceRecordDOJI {
    private final int id;
    private final String typeName;
    private final String company;
    private final String buyValue;
    private final String sellValue;

    public GoldPriceRecordDOJI(int id, String typeName, String company, String buyValue, String sellValue) {
        this.id = id;
        this.typeName = typeName;
        this.company = company;
        this.buyValue = buyValue;
        this.sellValue = sellValue;
    }

    public int getId() {
        return id;
    }

    public String getTypeName() {
        return typeName;
    }

    public String getCompany() {
        return company;
    }

    public String getBuyValue() {
        return buyValue;
    }

    public String getSellValue() {
        return sellValue;
    }

    public String toCsvLine() {
        return id + "," + typeName + "," + company + "," + buyValue + "," + sellValue;
    }
}
